File: lab03/calculadora_integral.go
package main

import (
	"fmt"
	"math"
	"runtime"
	"sync"
)

type CalculadoraIntegral struct {
	a                  float64
	b                  float64
	tolerancia         float64
	resultadoAnterior  float64
	numeroTrapecios    int
	nucleos            int
}

func NuevaCalculadora(a, b float64) *CalculadoraIntegral {
	return &CalculadoraIntegral{
		a:               a,
		b:               b,
		tolerancia:      0.0001,
		resultadoAnterior: 0,
		numeroTrapecios: 1,
		nucleos:         runtime.NumCPU(),
	}
}

func (c *CalculadoraIntegral) EvaluarFuncion(x float64) float64 {
	return 2*x*x + 3*x + 0.5
}

func (c *CalculadoraIntegral) CalcularIntegralConGoroutines() float64 {
	ancho := (c.b - c.a) / float64(c.numeroTrapecios)
	resultados := make([]float64, c.nucleos)
	var wg sync.WaitGroup

	trapeciosPorThread := c.numeroTrapecios / c.nucleos
	trapeciosRestantes := c.numeroTrapecios % c.nucleos

	wg.Add(c.nucleos)

	for i := 0; i < c.nucleos; i++ {
		go func(indice int) {
			defer wg.Done()

			inicio := indice * trapeciosPorThread
			fin := inicio + trapeciosPorThread
			if indice == c.nucleos-1 {
				fin += trapeciosRestantes
			}

			suma_local := 0.0
			for j := inicio; j < fin; j++ {
				x1 := c.a + float64(j)*ancho
				x2 := x1 + ancho
				altura := (c.EvaluarFuncion(x1) + c.EvaluarFuncion(x2)) / 2
				suma_local += altura * ancho
			}
			resultados[indice] = suma_local
		}(i)
	}

	wg.Wait()

	suma := 0.0
	for _, resultado := range resultados {
		suma += resultado
	}

	return suma
}

func (c *CalculadoraIntegral) CalcularIntegralConPoolWorkers() float64 {
	ancho := (c.b - c.a) / float64(c.numeroTrapecios)
	
	type Tarea struct {
		inicio int
		fin    int
		indice int
	}

	tareas := make(chan Tarea, c.nucleos)
	resultados := make([]float64, c.nucleos)
	var wg sync.WaitGroup

	trapeciosPorThread := c.numeroTrapecios / c.nucleos
	trapeciosRestantes := c.numeroTrapecios % c.nucleos

	for i := 0; i < c.nucleos; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for tarea := range tareas {
				suma_local := 0.0
				for j := tarea.inicio; j < tarea.fin; j++ {
					x1 := c.a + float64(j)*ancho
					x2 := x1 + ancho
					altura := (c.EvaluarFuncion(x1) + c.EvaluarFuncion(x2)) / 2
					suma_local += altura * ancho
				}
				resultados[tarea.indice] = suma_local
			}
		}()
	}

	for i := 0; i < c.nucleos; i++ {
		inicio := i * trapeciosPorThread
		fin := inicio + trapeciosPorThread
		if i == c.nucleos-1 {
			fin += trapeciosRestantes
		}
		tareas <- Tarea{inicio, fin, i}
	}

	close(tareas)
	wg.Wait()

	suma := 0.0
	for _, resultado := range resultados {
		suma += resultado
	}

	return suma
}

func (c *CalculadoraIntegral) SimularHastaConvergencia(usarPool bool) {
	metodo := "Goroutines Directas"
	if usarPool {
		metodo = "Pool de Workers"
	}

	fmt.Println("Iniciando cálculo de integral: f(x) = 2x² + 3x + 0.5")
	fmt.Printf("Intervalo: [%.0f, %.0f]\n", c.a, c.b)
	fmt.Printf("Método: %s\n", metodo)
	fmt.Printf("Núcleos disponibles: %d\n", c.nucleos)
	fmt.Println("-----------------------------------")

	for {
		var resultado float64
		if usarPool {
			resultado = c.CalcularIntegralConPoolWorkers()
		} else {
			resultado = c.CalcularIntegralConGoroutines()
		}

		diferencia := math.Abs(resultado - c.resultadoAnterior)
		fmt.Printf("Trapecios: %d | Resultado: %.10f | Diferencia: %.10f\n", 
			c.numeroTrapecios, resultado, diferencia)

		if c.numeroTrapecios > 1 && diferencia < c.tolerancia {
			fmt.Println("-----------------------------------")
			fmt.Println("Convergencia alcanzada!")
			fmt.Printf("Resultado final: %.10f\n", resultado)
			break
		}

		c.resultadoAnterior = resultado
		c.numeroTrapecios *= 2
	}
}

func main() {
	fmt.Println("=== USANDO GOROUTINES DIRECTAS ===\n")
	calculadora1 := NuevaCalculadora(2, 20)
	calculadora1.SimularHastaConvergencia(false)

	fmt.Println("\n\n=== USANDO POOL DE WORKERS ===\n")
	calculadora2 := NuevaCalculadora(2, 20)
	calculadora2.SimularHastaConvergencia(true)
}

File: lab03/CalculadoraIntegral.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <mutex>
#include <queue>
#include <condition_variable>
#include <functional>
#include <cmath>
#include <iomanip>

class CalculadoraIntegral {
public:
    double a;
    double b;
    double tolerancia;
    double resultadoAnterior;
    int numeroTrapecios;
    int nucleos;

    CalculadoraIntegral(double a, double b) 
        : a(a), b(b), tolerancia(0.0001), resultadoAnterior(0), 
          numeroTrapecios(1), nucleos(std::thread::hardware_concurrency()) {}

    double evaluarFuncion(double x) {
        return 2 * x * x + 3 * x + 0.5;
    }

    double calcularIntegralConThreads() {
        double ancho = (b - a) / numeroTrapecios;
        std::vector<double> resultados(nucleos, 0.0);
        std::vector<std::thread> threads;

        int trapeciosPorThread = numeroTrapecios / nucleos;
        int trapeciosRestantes = numeroTrapecios % nucleos;

        for (int i = 0; i < nucleos; i++) {
            int inicio = i * trapeciosPorThread;
            int fin = (i == nucleos - 1) ? inicio + trapeciosPorThread + trapeciosRestantes : inicio + trapeciosPorThread;
            
            threads.emplace_back([this, &resultados, i, inicio, fin, ancho]() {
                double suma_local = 0.0;
                for (int j = inicio; j < fin; j++) {
                    double x1 = a + j * ancho;
                    double x2 = x1 + ancho;
                    double altura = (evaluarFuncion(x1) + evaluarFuncion(x2)) / 2;
                    suma_local += altura * ancho;
                }
                resultados[i] = suma_local;
            });
        }

        for (auto& thread : threads) {
            thread.join();
        }

        double suma = 0.0;
        for (double resultado : resultados) {
            suma += resultado;
        }

        return suma;
    }
};

class PoolThreads {
private:
    std::vector<std::thread> workers;
    std::queue<std::function<void()>> tareas;
    std::mutex queueMutex;
    std::condition_variable cv;
    bool detener;

public:
    PoolThreads(int numWorkers) : detener(false) {
        for (int i = 0; i < numWorkers; i++) {
            workers.emplace_back([this]() {
                while (true) {
                    std::function<void()> tarea;
                    {
                        std::unique_lock<std::mutex> lock(queueMutex);
                        cv.wait(lock, [this]() { return !tareas.empty() || detener; });
                        
                        if (detener && tareas.empty()) {
                            return;
                        }
                        
                        if (!tareas.empty()) {
                            tarea = std::move(tareas.front());
                            tareas.pop();
                        }
                    }
                    if (tarea) {
                        tarea();
                    }
                }
            });
        }
    }

    template<class F>
    void encolaTarea(F f) {
        {
            std::unique_lock<std::mutex> lock(queueMutex);
            tareas.emplace(f);
        }
        cv.notify_one();
    }

    ~PoolThreads() {
        {
            std::unique_lock<std::mutex> lock(queueMutex);
            detener = true;
        }
        cv.notify_all();
        for (auto& worker : workers) {
            worker.join();
        }
    }
};

class CalculadoraConPool {
private:
    double a;
    double b;
    double tolerancia;
    double resultadoAnterior;
    int numeroTrapecios;
    int nucleos;

public:
    CalculadoraConPool(double a, double b)
        : a(a), b(b), tolerancia(0.0001), resultadoAnterior(0),
          numeroTrapecios(1), nucleos(std::thread::hardware_concurrency()) {}

    double evaluarFuncion(double x) {
        return 2 * x * x + 3 * x + 0.5;
    }

    double calcularIntegralConPoolThreads() {
        double ancho = (b - a) / numeroTrapecios;
        std::vector<double> resultados(nucleos, 0.0);
        PoolThreads pool(nucleos);

        int trapeciosPorThread = numeroTrapecios / nucleos;
        int trapeciosRestantes = numeroTrapecios % nucleos;

        std::mutex resultMutex;
        int tareasCompletadas = 0;
        std::condition_variable allDone;

        for (int i = 0; i < nucleos; i++) {
            int inicio = i * trapeciosPorThread;
            int fin = (i == nucleos - 1) ? inicio + trapeciosPorThread + trapeciosRestantes : inicio + trapeciosPorThread;
            
            pool.encolaTarea([this, &resultados, &resultMutex, &tareasCompletadas, &allDone, i, inicio, fin, ancho]() {
                double suma_local = 0.0;
                for (int j = inicio; j < fin; j++) {
                    double x1 = a + j * ancho;
                    double x2 = x1 + ancho;
                    double altura = (evaluarFuncion(x1) + evaluarFuncion(x2)) / 2;
                    suma_local += altura * ancho;
                }
                {
                    std::unique_lock<std::mutex> lock(resultMutex);
                    resultados[i] = suma_local;
                    tareasCompletadas++;
                }
                allDone.notify_all();
            });
        }

        {
            std::unique_lock<std::mutex> lock(resultMutex);
            allDone.wait(lock, [&tareasCompletadas, this]() { return tareasCompletadas == nucleos; });
        }

        double suma = 0.0;
        for (double resultado : resultados) {
            suma += resultado;
        }

        return suma;
    }

    void simularHastaConvergencia(bool usarPool) {
        std::string metodo = usarPool ? "Pool de Threads" : "Threads Directos";
        
        std::cout << "Iniciando calculo de integral: f(x) = 2x2 + 3x + 0.5" << std::endl;
        std::cout << "Intervalo: [" << a << ", " << b << "]" << std::endl;
        std::cout << "Metodo: " << metodo << std::endl;
        std::cout << "Nucleos disponibles: " << nucleos << std::endl;
        std::cout << "-----------------------------------" << std::endl;

        CalculadoraIntegral calc1(a, b);
        
        while (true) {
            double resultado = usarPool ? calcularIntegralConPoolThreads() : calc1.calcularIntegralConThreads();
            double diferencia = std::abs(resultado - resultadoAnterior);

            std::cout << "Trapecios: " << numeroTrapecios << " | Resultado: " 
                      << std::fixed << std::setprecision(10) << resultado 
                      << " | Diferencia: " << diferencia << std::endl;

            if (numeroTrapecios > 1 && diferencia < tolerancia) {
                std::cout << "-----------------------------------" << std::endl;
                std::cout << "Convergencia alcanzada!" << std::endl;
                std::cout << "Resultado final: " << std::fixed << std::setprecision(10) << resultado << std::endl;
                break;
            }

            resultadoAnterior = resultado;
            numeroTrapecios *= 2;
            calc1.numeroTrapecios = numeroTrapecios;
        }
    }
};

int main() {
    std::cout << "=== USANDO THREADS DIRECTOS ===" << std::endl << std::endl;
    CalculadoraConPool calc1(2, 20);
    calc1.simularHastaConvergencia(false);

    std::cout << std::endl << std::endl;
    std::cout << "=== USANDO POOL DE THREADS ===" << std::endl << std::endl;
    CalculadoraConPool calc2(2, 20);
    calc2.simularHastaConvergencia(true);

    return 0;
}
